### ISSUE TITLE
perlPackages.ModuleBuild (a dependency of git) fails to cross-compile

### ISSUE THREAD

**Issue by lopsided98** [NixOS/nixpkgs#66741]
`perlPackages.ModuleBuild` fails to cross-compile, which prevents `git` from cross-compiling (unless `perlSupport = false`). I don't have much experience with Perl, but my understanding is that `Scalar::Util` is trying to load an XS module. I'm not sure how to work around this.

```
unpacking sources
unpacking source archive /nix/store/jwpgfd4pziwrrx2hywr9ckf9xil7aizc-Module-Build-0.4229.tar.gz
source root is Module-Build-0.4229
setting SOURCE_DATE_EPOCH to timestamp 1555368693 of file Module-Build-0.4229/t/xs.t
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
patching ./t/properties/share_dir.t...
patching ./t/compat/exit.t...
patching ./t/xs.t...
patching ./t/write_default_maniskip.t...
patching ./t/versions.t...
patching ./t/use_tap_harness.t...
patching ./t/unit_run_test_harness.t...
patching ./t/tilde.t...
patching ./t/test_types.t...
patching ./t/test_type.t...
patching ./t/test_reqs.t...
patching ./t/test_file_exts.t...
patching ./t/signature.t...
patching ./t/script_dist.t...
patching ./t/runthrough.t...
patching ./t/ppm.t...
patching ./t/pod_parser.t...
patching ./t/parents.t...
patching ./t/par.t...
patching ./t/notes.t...
patching ./t/new_from_context.t...
patching ./t/mymeta.t...
patching ./t/metadata2.t...
patching ./t/metadata.t...
patching ./t/manifypods.t...
patching ./t/install_extra_target.t...
patching ./t/install.t...
patching ./t/help.t...
patching ./t/files.t...
patching ./t/extend.t...
patching ./t/ext.t...
patching ./t/destinations.t...
patching ./t/debug.t...
patching ./t/compat.t...
patching ./t/basic.t...
patching ./t/add_property_hash.t...
patching ./t/add_property_array.t...
patching ./t/add_property.t...
patching ./t/PL_files.t...
patching ./bin/config_data...
# running Build.PL --prefix /nix/store/9lm1ijj01kmimdplwrs77faflwzxjir3-perl5.28.2-Module-Build-0.4229-armv6l-unknown-linux-gnueabihf --installdirs site --config perl=/nix/store/dv4w3l6jclic0wqbkpwv6m79wbfkvk9n-perl-5.28.2-armv6l-unknown-linux-gnueabihf-dev/bin/perl --fullperl "/nix/store/by0xi2dd0aha8armhbkd2a3zdqzrjnmi-perl-5.28.2/bin/perl"
Checking prerequisites...
  test_requires:
    !  Couldn't find a $VERSION in prerequisite File::Temp

Checking optional features...
inc_bundling_support....disabled
  requires:
    ! inc::latest is not installed

ERRORS/WARNINGS FOUND IN PREREQUISITES.  You may wish to install the versions
of the modules indicated above before proceeding with this installation

Attempt to reload Scalar/Util.pm aborted.
Compilation failed in require at /nix/store/81lwy2awhy4fqxxiyqcc9jflslrzl2bm-perl-5.28.2-armv6l-unknown-linux-gnueabihf/lib/perl5/5.28.2/armv6l-linux/Data/Dumper.pm line 312.
Couldn't run Build.PL: No such file or directory at lib/Module/Build/Compat.pm line 341.
```

Maintainer information:
```yaml
attribute:
  - perlPackages.ModuleBuild
  - git
```

cc @shlevy @volth because you've worked on Perl cross-compiling

**Comment by volth** [comment]
> but my understanding is that `Scalar::Util` is trying to load an XS module.

Yes, it is tricky to cross-compile Perl packages with C code.
Those which do work (`LWP`, `XMLParser`, ...) require custom patches

**Comment by lopsided98** [comment]
I read through the explanation in the manual about creating a stub implementation, but I'm not sure what to do in this case because `Scalar::Util` is part of Perl core.

**Comment by volth** [comment]
I tried to fix cross-compilation of `perlPackages.ModuleBuild` few months ago but failed (of course it is feasible, but not easy), so I limited then with fixing of only packages which are used by `switch-to-configuration.pl` in order to be able to cross-compile NixOS system closures for ARM.

**Comment by lopsided98** [comment]
Personally, I don't really care about `ModuleBuild`, I just want `git` to cross-compile out of the box. Maybe the best solution is just to set `perlSupport = false` when cross-compiling, or disable `git-instaweb`.

**Comment by volth** [comment]
maybe
```diff
- perlSupport ? true
+ perlSupport ? stdenv.buildPlatform == stdenv.hostPlatform
```

**Comment by lopsided98** [comment]
I'm pretty sure this hasn't been fixed.

**Comment by yorickvP** [comment]
Still happening

**Comment by lopsided98** [comment]
This now breaks cross-compiled NixOS because `perlPackages.XMLParser` transitively depends on `ModuleBuild` after https://github.com/NixOS/nixpkgs/commit/a0d34c753b85e8207510f0e4d9001906250973e4.

**Comment by stigtsp** [comment]
Hey. I think reverting a0d34c7 will be a reasonable short term fix if it blocking cross-compilation. With it reverted, XML::Parser builds:

```
nix-build -A pkgsCross.aarch64-multiplatform.perldevelPackages.XMLParser
```

**Comment by lopsided98** [comment]
Yeah, reverting that commit won't fix git because `ModuleBuild` is still broken, it would just stop `ModuleBuild` from being a requirement to build a NixOS system.

**Comment by lopsided98** [comment]
`ModuleBuild` still fails to cross-compile, but Perl support is git is now disabled when cross-compiling.

**Comment by ju1m** [comment]
This also prevents `xdg-utils` to cross-compile. And thus `services.x11.xserver.enable = true` fails to cross-compile.

**Comment by zakame** [comment]
Yeah, its basically the same thing as #115170 - the failing step

```
Checking prerequisites...
  test_requires:
    !  Couldn't find a $VERSION in prerequisite File::Temp
```

Is due to https://metacpan.org/release/LEONT/Module-Build-0.4231/source/Build.PL#L67 which doesn't really make sense for our purposes as `File::Temp` in our Perls are now way past 0.15:

```perl
  test_requires         => {
    'File::Temp'            => 0.15,    # tmpdir() + fixes
    'Test::More'            => 0.49,
```

```shell
$ nix-shell -p perl --run 'perl -MFile::Temp -E "say \$File::Temp::VERSION"'
0.2311
```

`postPatch` PR to follow shortly.

**Comment by ius** [comment]
This once again breaks cross compiled NixOS because of https://github.com/NixOS/nixpkgs/pull/154809 pulling this in via `ConfigIniFiles`

**Comment by dasJ** [comment]
stc sould be fixed by #158021

**Comment by ius** [comment]
_Flips sign_: 0 days since the last incident. NixOS fails to cross compile again.

This time it's https://github.com/NixOS/nixpkgs/pull/157329 depending on _Array-Compare_ which pulls in _Moo_, which lists _ModuleRuntime_ in _propagatedBuildInputs_. As long as there's no structural fix for this issue I'm afraid we risk running into it on every rework of the core perl scripts..

@dasJ If you happen to have some time to look at it again, or someone else who _does_ know perl...

**Comment by stigtsp** [comment]
Attempted to do a quick fix by looking more closely at modules pulled in by `Array::Compare` and it's dependencies, but it quickly becomes a rabbit hole where `Module::Build` or a dynamic perl (in case of tests for Type::Tiny) is needed. Need figure out how to fix Module::Build for cross builds to solve this class of recurring problems ...

Anyway, I've added a workaround for stc that avoids Array::Compare here #160176

**Comment by lopsided98** [comment]
This particular issue has been fixed by #225640, although Git with Perl support still fails due to a different error in `libnet`:
```
Can't load module IO, dynamic loading not available in this perl.
  (You may need to build a new perl executable which either supports
  dynamic loading or has the IO module statically linked into it.)
 at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/Handle.pm line 268.
Compilation failed in require at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/Handle.pm line %.
BEGIN failed--compilation aborted at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/Handle.pm line %.
Compilation failed in require at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/Seekable.pm line %.
BEGIN failed--compilation aborted at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/Seekable.pm line %.
Compilation failed in require at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/File.pm line %.
BEGIN failed--compilation aborted at /nix/store/1yicwrwa8izkvf9vpv93wsznx5iq5pf4-perl-armv6l-unknown-linux-gnueabihf-5.38.0/lib/perl5/5.38.0/armv6l-linux/IO/File.pm line %.
Compilation failed in require at Configure line %.
BEGIN failed--compilation aborted at Configure line %.
Generating a Unix-style Makefile
Writing Makefile for Net
no configure script, doing nothing
building
build flags: SHELL=/nix/store/9b4skch97sv02rwmczbxp1pwg382hv84-bash-5.2-p15/bin/bash
make: *** No rule to make target 'libnet.cfg', needed by 'pm_to_blib'.  Stop.
```
